File: agenttel-agent/src/test/java/io/agenttel/agent/mcp/McpServerTest.java
package io.agenttel.agent.mcp;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.*;

class McpServerTest {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private McpServer server;
    private int port;
    private HttpClient httpClient;

    @BeforeEach
    void setUp() throws Exception {
        port = 18081 + (int) (Math.random() * 1000);
        server = new McpServer(port);
        server.registerTool(
                new McpToolDefinition(
                        "test_tool",
                        "A test tool",
                        Map.of("input", new McpToolDefinition.ParameterDefinition("string", "Test input")),
                        List.of("input")
                ),
                args -> "Result: " + args.get("input")
        );
        server.start();
        httpClient = HttpClient.newHttpClient();
    }

    @AfterEach
    void tearDown() {
        server.stop();
    }

    @Test
    void healthCheck_returnsOk() throws Exception {
        var response = httpClient.send(
                HttpRequest.newBuilder()
                        .uri(URI.create("http://localhost:" + port + "/health"))
                        .GET()
                        .build(),
                HttpResponse.BodyHandlers.ofString());

        assertThat(response.statusCode()).isEqualTo(200);
        JsonNode body = MAPPER.readTree(response.body());
        assertThat(body.get("status").asText()).isEqualTo("ok");
        assertThat(body.get("tools").asInt()).isEqualTo(1);
    }

    @Test
    void initialize_returnsProtocolVersion() throws Exception {
        String request = MAPPER.writeValueAsString(Map.of(
                "jsonrpc", "2.0", "id", 1, "method", "initialize", "params", Map.of()));

        var response = httpClient.send(
                HttpRequest.newBuilder()
                        .uri(URI.create("http://localhost:" + port + "/mcp"))
                        .POST(HttpRequest.BodyPublishers.ofString(request))
                        .header("Content-Type", "application/json")
                        .build(),
                HttpResponse.BodyHandlers.ofString());

        assertThat(response.statusCode()).isEqualTo(200);
        JsonNode body = MAPPER.readTree(response.body());
        assertThat(body.get("result").get("protocolVersion").asText()).isEqualTo("2024-11-05");
        assertThat(body.get("result").get("serverInfo").get("name").asText()).isEqualTo("agenttel-mcp");
    }

    @Test
    void toolsList_returnsRegisteredTools() throws Exception {
        String request = MAPPER.writeValueAsString(Map.of(
                "jsonrpc", "2.0", "id", 1, "method", "tools/list", "params", Map.of()));

        var response = httpClient.send(
                HttpRequest.newBuilder()
                        .uri(URI.create("http://localhost:" + port + "/mcp"))
                        .POST(HttpRequest.BodyPublishers.ofString(request))
                        .header("Content-Type", "application/json")
                        .build(),
                HttpResponse.BodyHandlers.ofString());

        JsonNode body = MAPPER.readTree(response.body());
        JsonNode tools = body.get("result").get("tools");
        assertThat(tools.isArray()).isTrue();
        assertThat(tools.size()).isEqualTo(1);
        assertThat(tools.get(0).get("name").asText()).isEqualTo("test_tool");
    }

    @Test
    void toolsCall_executesTool() throws Exception {
        String request = MAPPER.writeValueAsString(Map.of(
                "jsonrpc", "2.0", "id", 1, "method", "tools/call",
                "params", Map.of("name", "test_tool", "arguments", Map.of("input", "hello"))));

        var response = httpClient.send(
                HttpRequest.newBuilder()
                        .uri(URI.create("http://localhost:" + port + "/mcp"))
                        .POST(HttpRequest.BodyPublishers.ofString(request))
                        .header("Content-Type", "application/json")
                        .build(),
                HttpResponse.BodyHandlers.ofString());

        JsonNode body = MAPPER.readTree(response.body());
        JsonNode content = body.get("result").get("content");
        assertThat(content.get(0).get("text").asText()).isEqualTo("Result: hello");
    }

    @Test
    void toolsCall_unknownTool_returnsError() throws Exception {
        String request = MAPPER.writeValueAsString(Map.of(
                "jsonrpc", "2.0", "id", 1, "method", "tools/call",
                "params", Map.of("name", "nonexistent")));

        var response = httpClient.send(
                HttpRequest.newBuilder()
                        .uri(URI.create("http://localhost:" + port + "/mcp"))
                        .POST(HttpRequest.BodyPublishers.ofString(request))
                        .header("Content-Type", "application/json")
                        .build(),
                HttpResponse.BodyHandlers.ofString());

        JsonNode body = MAPPER.readTree(response.body());
        assertThat(body.has("error")).isTrue();
        assertThat(body.get("error").get("message").asText()).contains("Unknown tool");
    }

    @Test
    void unknownMethod_returnsMethodNotFoundError() throws Exception {
        String request = MAPPER.writeValueAsString(Map.of(
                "jsonrpc", "2.0", "id", 1, "method", "nonexistent"));

        var response = httpClient.send(
                HttpRequest.newBuilder()
                        .uri(URI.create("http://localhost:" + port + "/mcp"))
                        .POST(HttpRequest.BodyPublishers.ofString(request))
                        .header("Content-Type", "application/json")
                        .build(),
                HttpResponse.BodyHandlers.ofString());

        JsonNode body = MAPPER.readTree(response.body());
        assertThat(body.has("error")).isTrue();
        assertThat(body.get("error").get("code").asInt()).isEqualTo(-32601);
    }
}

File: agenttel-node/tests/mcp-server.test.ts
import { describe, it, expect } from 'vitest';
import { MCPServer } from '../src/agent/mcp/server.js';

describe('MCPServer', () => {
  it('returns initialize response with protocol version and capabilities', async () => {
    const server = new MCPServer('agenttel-test', '1.0.0');

    const resp = await server.handleRequest(
      JSON.stringify({ jsonrpc: '2.0', id: 1, method: 'initialize' }),
    );
    const parsed = JSON.parse(resp);

    expect(parsed.jsonrpc).toBe('2.0');
    expect(parsed.id).toBe(1);
    expect(parsed.error).toBeUndefined();
    expect(parsed.result.protocolVersion).toBe('2024-11-05');
    expect(parsed.result.capabilities).toEqual({ tools: {} });
    expect(parsed.result.serverInfo).toEqual({ name: 'agenttel-test', version: '1.0.0' });
  });

  it('returns empty tools list', async () => {
    const server = new MCPServer('test', '1.0');

    const resp = await server.handleRequest(
      JSON.stringify({ jsonrpc: '2.0', id: 1, method: 'tools/list' }),
    );
    const parsed = JSON.parse(resp);

    expect(parsed.result.tools).toEqual([]);
  });

  it('returns registered tools in tools/list', async () => {
    const server = new MCPServer('test', '1.0');

    server.registerTool(
      { name: 'get_health', description: 'Get service health', inputSchema: { type: 'object' } },
      async () => ({ status: 'ok' }),
    );
    server.registerTool(
      { name: 'get_metrics', description: 'Get metrics', inputSchema: { type: 'object' } },
      async () => ({}),
    );

    const resp = await server.handleRequest(
      JSON.stringify({ jsonrpc: '2.0', id: 1, method: 'tools/list' }),
    );
    const parsed = JSON.parse(resp);

    expect(parsed.result.tools).toHaveLength(2);
    const names = parsed.result.tools.map((t: any) => t.name);
    expect(names).toContain('get_health');
    expect(names).toContain('get_metrics');
  });

  it('calls tool and returns result', async () => {
    const server = new MCPServer('test', '1.0');

    server.registerTool(
      { name: 'greet', description: 'Greet someone', inputSchema: { type: 'object' } },
      async (params) => ({ greeting: `Hello, ${params.name}` }),
    );

    const resp = await server.handleRequest(
      JSON.stringify({
        jsonrpc: '2.0',
        id: 1,
        method: 'tools/call',
        params: { name: 'greet', arguments: { name: 'Alice' } },
      }),
    );
    const parsed = JSON.parse(resp);

    expect(parsed.error).toBeUndefined();
    expect(parsed.result.content).toHaveLength(1);
    expect(parsed.result.content[0].type).toBe('text');

    const text = JSON.parse(parsed.result.content[0].text);
    expect(text.greeting).toBe('Hello, Alice');
  });

  it('returns error for unknown tool', async () => {
    const server = new MCPServer('test', '1.0');

    const resp = await server.handleRequest(
      JSON.stringify({
        jsonrpc: '2.0',
        id: 1,
        method: 'tools/call',
        params: { name: 'nonexistent', arguments: {} },
      }),
    );
    const parsed = JSON.parse(resp);

    expect(parsed.error).toBeDefined();
    expect(parsed.error.code).toBe(-32602);
    expect(parsed.error.message).toContain('Unknown tool');
  });

  it('returns error for unknown method', async () => {
    const server = new MCPServer('test', '1.0');

    const resp = await server.handleRequest(
      JSON.stringify({ jsonrpc: '2.0', id: 1, method: 'unknown/method' }),
    );
    const parsed = JSON.parse(resp);

    expect(parsed.error).toBeDefined();
    expect(parsed.error.code).toBe(-32601);
    expect(parsed.error.message).toContain('Method not found');
  });

  it('returns parse error for invalid JSON', async () => {
    const server = new MCPServer('test', '1.0');

    const resp = await server.handleRequest('{invalid json');
    const parsed = JSON.parse(resp);

    expect(parsed.error).toBeDefined();
    expect(parsed.error.code).toBe(-32700);
    expect(parsed.error.message).toBe('Parse error');
  });
});
